docs(parameters): add DevOps Dojo: ROS Node Config documentation

kaspermeck-arm opened the pull request:
Updated the guidelines for how to configure the ROS node parameters.

## Description

Updating coding guidelines for the ROS node parameters according to the decision made in _DevOps Dojo: Ros Node Config _ see [DevOps Dojo: ROS Node Configuration - Update Parameter Contribution Guidelines](https://github.com/autowarefoundation/autoware-documentation/issues/352)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

kaspermeck-arm commented:
I noticed that _pre-commit.ci_ made some changes
* https://github.com/autowarefoundation/autoware-documentation/pull/367/commits/ae75f5f1a3bcd922ad401a905155bc289b90e351

I think this decreases readability and shouldn't be used for the JSON schema either. Is this how we want it to be implemented?

kminoda commented:
Thank you for your contribution! 

What is the current status of this PR? Is it ready for review? 
Given that there is some comment like `(Original content, will need updating)` in the PR, and also some required CI is failing, is this PR still not ready for review?

xmfcx commented:
@kaspermeck-arm thanks for the PR!
- I've changed the PR title to fix the semantic-pull-request CI test.
- Could you also fix the https://results.pre-commit.ci/run/github/429694350/1684157254.Y4cUa-XoSEuJwgn5B6SUBA pre-commit.ci checks too?
- I will mark this as draft until they are fixed, you can set it as "Ready for review" once the Required checks pass.
  - You can ignore the DCO one for now.

xmfcx commented:
> I noticed that _pre-commit.ci_ made some changes
> 
>     * [ae75f5f](https://github.com/autowarefoundation/autoware-documentation/commit/ae75f5f1a3bcd922ad401a905155bc289b90e351)
> 
> 
> I think this decreases readability and shouldn't be used for the JSON schema either. Is this how we want it to be implemented?

It refactors them based on length.
I think for small number of items, this is fine.
For large number of items, it will turn them into multi-line lists anyway. So we don't need to take an action for this.

kaspermeck-arm commented:
@kminoda, thanks for your comments!
We're currently working on the next DevOps Dojo which will address the ROS Node Launch files, so the comment there acts as a placeholder. Once the dojo is complete, I'll update this section. Is this approach OK?

@xmfcx, thanks for your comments!
I rebased the commit and updated it to follow the comments from `markdownlint`.

kaspermeck-arm commented:
@kminoda @xmfcx

It seems as though some links are broken in some other [files](https://github.com/autowarefoundation/autoware-documentation/actions/runs/4994340506/jobs/8944761037?pr=367). What's the process of resolving things?

Thanks for your patience while I learn how Autoware PRs work!

xmfcx commented:
> It seems as though some links are broken in some other [files](https://github.com/autowarefoundation/autoware-documentation/actions/runs/4994340506/jobs/8944761037?pr=367). What's the process of resolving things?

You can ignore them, they can be fixed later. Right now all `Required` CI tests are passed.

> Thanks for your patience while I learn how Autoware PRs work!

Thanks for working on this, I will review as soon as possible.

kaspermeck-arm commented:
@xmfcx @kminoda 

Any chance this can get merged before our next Open AD Kit meeting on May 25th? Thanks!

xmfcx commented on docs/contributing/coding-guidelines/ros-nodes/parameters.md:
```diff
@@ -1,36 +1,150 @@
+- [About ROS 2 Parameters](https://docs.ros.org/en/humble/Concepts/About-ROS-2-Parameters.html)
 
-Autoware has the following two types of parameter files for ROS packages:
+## JSON Schema
```
As noted in [our previous discussion](https://github.com/orgs/autowarefoundation/discussions/3325#discussioncomment-5417178), we came to an agreement on locating the JSON Schema files in the `package/schema/default_node_name.schema.json` path.

However, it's important to acknowledge that Picknik's library doesn't incorporate this feature yet. And our CI actions are not yet fully prepared to accommodate the automatic association feature of VS Code through the `.vscode/settings.json` file. This is an aspect we all are aware of and should work towards.

Because of this, I'd propose that we consider introducing JSON Schema related documentation only after resolving these two hurdles. This approach would ensure we're not putting the cart before the horse, thus preventing potential confusion amongst our developers regarding the actual functionality of these JSON Schema files.

That being said, it could be beneficial to outline our prospective plans for the use of JSON Schema files in this documentation. This would provide a clear roadmap for the Autoware development team, giving us a sense of direction and a better understanding of the upcoming improvements to our workflow.

kaspermeck-arm replied:
@xmfcx @mitsudome-r @HamburgDave @oguzkaganozt 

The content of the contribution guidelines has already been approved, see https://github.com/orgs/autowarefoundation/discussions/3371#discussioncomment-5544013. 

Regarding the location of the JSON Schema, that's what the contribution guidelines state. From the commit:

> The schema file path is `INSERT_PATH_TO_PACKAGE/schema/` and the schema file name is `INSERT_NODE_NAME.schema.json`. To adapt the template to the ROS node, replace each `INSERT_...` and add all parameters `1..N`.

Or did I miss something?

xmfcx replied:
I appreciate the ongoing discussion and your continuous engagement in shaping our contribution guidelines, as per the shared agreement in [this thread](https://github.com/orgs/autowarefoundation/discussions/3371#discussioncomment-5544013).

The conversations have been extensive, and it's important to revisit and consolidate our shared understanding.

Our primary aim from the outset has been to establish a single source of truth, with a focus on omitting default values from the nodes. To that end, we've integrated ideas from both @kenji-miyake's [generate_parameter_library](https://github.com/PickNikRobotics/generate_parameter_library) and @doganulus' JSON Schema files, despite initial differing perspectives.

Our current consensus, as I understand, is to provide the flexibility to choose between using: 
- the 'generate_parameter_library'
- or standard ROS 2 parameters, provided the defaults are not set in the source files.

Furthermore, once the 'generate_parameter_library' can read from JSON Schema files, these files can be included in the schema folder.

[Your video from the DevOps Dojo: ROS Node Configuration - Final3 Proposal](https://user-images.githubusercontent.com/30662532/228092478-1a61b9d4-1109-40a5-973c-773eb53ac40b.mp4) illustrates the current requirement for manual input of the `.schema.json` file path.
![Screenshot from 2023-05-24 18-07-14](https://github.com/autowarefoundation/autoware-documentation/assets/10751153/ab784f5d-6abb-4c1d-87c5-c032187fc8b5)
- Here, you've made VS Code attempt to validate all YAML Param files (with `*.param.yaml`)
- If you try to open any other YAML file, it will be shown as invalidated
- For this to work correctly, user also needs to enter the full path of the `.yaml` file or files which this schema is trying to validate.

For enhanced user experience, I propose that we create a CI action to generate the `settings.json` file from the schema files. This way, users wouldn't need to manually specify these details, reducing the likelihood of typing errors. Once this is accomplished, the next step would be to validate the `.yaml` files using CI, based on the schema files.

Only after we've taken these steps should we consider adding this behavior to the documentation and recommending developers to adopt it. Doing so prematurely may lead to confusion without substantial benefit.

kaspermeck-arm replied:
There was a lot of discussion whether or not use PickNikRobotics library and in the end, it did not make the final proposal. However, the final proposal doesn't close the door on using PickNikRobotics library further down the road.

We decided that even though we don't utilize the JSON Schema immediately, creating it when doing the other work on the ROS nodes made sense and was decided on (without any objections) in the Open AD Kit meeting on April 6th. I pinged other AWF channels and individuals weeks ahead to get input.

Without the JSON Schema file you cannot do any validation regardless. It's a chicken and egg problem. An CI action can be created at any time and the work is outlined here:
- https://github.com/autowarefoundation/autoware-github-actions/issues/232

Create an auto-generated `settings.json` file:
- https://github.com/autowarefoundation/autoware/issues/3424

The JSON Schema is used in VSCode to help a developer understand what parameters are required for a node. To use VSCode is not obligatory and therefore it is under "Tips and Tricks", i.e., not part of the requirement, but as a tool to help our developers.

Can we please merge the new coding guidelines which have already been approved?

xmfcx replied:
I appreciate your commitment to refining our coding guidelines, and the rationale you've provided makes sense. However, I would like to express some concerns that arise from our collective discussions.

While we had multiple threads 
- [ROS Node Configuration](https://github.com/orgs/autowarefoundation/discussions/3260),
- [ROS Node Configuration - Final Proposal](https://github.com/orgs/autowarefoundation/discussions/3288),
- [ROS Node Configuration - Final2 Proposal](https://github.com/orgs/autowarefoundation/discussions/3325), 
- [DevOps Dojo: ROS Node Configuration - Final3 Proposal](https://github.com/orgs/autowarefoundation/discussions/3371),

my last comment in the Final2 Proposal [here](https://github.com/orgs/autowarefoundation/discussions/3325#discussioncomment-5417178) led to a unified conclusion from the remaining participants, but a new thread was then started which saw diminished engagement. This raises questions about the representation of viewpoints in the final proposal.

In the Open AD Kit meeting on April 6th, the decision to create JSON Schema files, despite their immediate non-utilization, was passed without objections. I acknowledge the effort you made to gather input weeks ahead.

The process this proposal underwent isn't as important to me as its content and potential impact on the Autoware repository. An uncontested approval should not deter further discussion on an idea, particularly one of this magnitude.

Regarding the 'chicken and egg' analogy, my perspective differs. I view this situation as more akin to 'putting the cart before the horse'. The initial schema file ought to be created concurrently with a supporting CI action. Currently, we lack a mechanism that validates the schema files themselves, and manually entering full paths for each file on VS Code is cumbersome.

In the documentation, we can indeed outline future plans about this feature, but as it stands, the existing provisions are insufficient for productive use. A more practical approach is needed before we can integrate these guidelines into our workflow.

We can discuss this in the next Open AD Kit meeting or raise it to the next TSC meeting if necessary.

doganulus replied:
@xmfcx I think a supporting CI action for schema automation is nice but I suggest not stating it as a blocking item. It is not that hard to enter full paths for each node by hand. Considering node names are not changing frequently, this is probably a one-time effort for each node. So even it may not need such automation initially. Automation may be needed to publish Autoware schemas at a remote endpoint as many cloud-native software do. 

One thing I want from the software group or the TSC is to standardize/canonize the paths of nodes across Autoware or let me know if such a decision is already in place. If you can give a guarantee to have paths like below and ensure it across Autoware,

```
`<prefix>`/`<component>`/`<package>`/{`include`, `src`, `config`, `schema`, `tests`}
```

Then most Autoware developers would not need to touch `vscode` settings once again. We need some conventions to keep it ordered and systematic.

xmfcx replied:
@doganulus thank you for your response. While I agree that it's possible for us to manually input full paths, it's important to remember that each package can potentially have multiple nodes, and each node can carry multiple parameter files.
**Examples:**
 - https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/config/dynamic_avoidance/dynamic_avoidance.param.yaml
 - https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/config/lane_change/lane_change.param.yaml

Considering this, the manual process can become significantly cumbersome and raise the maintenance cost.

It's true that node names don't frequently change, yet we should be mindful that changes can and do occur. Keeping the `settings.json` file updated in light of these changes would necessitate an ongoing maintenance effort, potentially diverting our attention from core development tasks.

In terms of standardizing the paths, I agree that conventions are crucial for maintaining an orderly and systematic workflow. We can potentially standardize paths for main components such as "planning, sensing, perception, etc.". However, given that nodes and packages are regularly updated and modified, the primary focus should be on developing a CI action that can create the `settings.json` file automatically.

With the absence of a validation process for `.schema.json` files currently, I strongly believe that having the automation in place before we advance is key to maintaining a manageable workload and a smooth workflow. Hence, I see this as a blocking item. And that automation should precede further advancement.

doganulus replied:
@xmfcx What do you expect from the `.schema.json` validation? Do you want to check the `schema.json` files to satisfy, say, the `draft-07` specification?

xmfcx replied:
> What do you expect from the .schema.json validation? Do you want to check the schema.json files to satisfy, say, the draft-07 specification?

@doganulus, you are correct. At its core, I would indeed want the `.schema.json` files to conform to specifications such as the `draft-07`, which is an established and reliable standard for JSON Schema.

Currently, we can manually check these files using tools like VS Code or potentially some online validators. However, this manual process isn't scalable or efficient in the long run.

Moreover, I envision a system that not only validates the `.schema.json` files but also employs these schemas to automatically validate the `.yaml` files. Such a system would greatly enhance our efficiency and reliability, ultimately reducing the likelihood of errors slipping through manual checks. Automation here serves a dual purpose - standardization and quality assurance.

isamu-takagi replied:
Hi, I just noticed this now, has it changed from the conclusion before? My understanding is:
- Creating a parameter definition file is recommended (but not required).
- When creating a parameter definition file, follow either the JSON schema or PickNik's generate_parameter_library format.

xmfcx replied:
> Hi, I just noticed this now, has it changed from the conclusion before? My understanding is:
> - Creating a parameter definition file is recommended (but not required).

@isamu-takagi this is correct.

> - When creating a parameter definition file, follow either the `JSON schema` or PickNik's `generate_parameter_library` format.

As for your second point, it's partially accurate. While we recommend following either the JSON Schema or PickNik's `generate_parameter_library` format, we must consider the current limitations. Implementing `.schema.json` files right now, without the necessary supporting mechanisms, could lead to additional maintenance efforts with limited advantages.

Therefore, I propose that we proceed with adding .schema.json files only when:

1. CI actions are in place to:
  - Validate the schema files (essential)
  - Validate the .yaml files with the schemas (essential)
  - Create settings.json files to support VS Code (desirable but not critical)
2. The `generate_parameter_library` can accommodate JSON Schemas, ensuring we do not end up with multiple sources of truth (desirable but not critical)

Without these capabilities, we risk having superfluous schema files in the repository, functional only when manually added to the VS Code's `settings.json` file.

In light of these considerations, I suggest that this proposal be implemented concurrently with the above-mentioned functionalities to maximize its effectiveness and efficiency.

isamu-takagi replied:
@xmfcx Sorry, I didn't understand about the validation and settings.json of VS Code. I think we are trying to do something more complicated than originally thought. For the time being, parameter definition file is a kind of README. Developers write parameter definition file instead of writing the parameters section of README.

So there is no particular difference between formats, maintenance costs are paid by the person who chooses the format. The only thing that matters to anyone other than the developer of that package is reading the parameter definition file. Of course, definition files are more convenient for those who want their tools to process the parameters.

isamu-takagi replied:
Or does someone other than the package maintainer add parameter definition files?

mitsudome-r replied:
@xmfcx 
I'm with @isamu-takagi -san on this.

The original reason that Kasper wanted to include JSON schema was to have a single source of truth about parameter definition. My understanding is that we have come to agreement in the GitHub Discussion to use JSON schema on that, and I think having that consensus defined in coding-guideline could happen regardless of the the existence of CI. Also, regarding the parameter generation using PickNick was decided to be optional thing that is up to the user whether they want to use it or not.

Is your concern that you don't want to have people creating JSON schema file that might not be valid by creating this PR public? In that case, we might be a valid concern and we can consider CI with the GitHub Actions, but I'm not sure why you want to have parameter generation feature available before having this guideline merged.

xmfcx replied:
@mitsudome-r said:
> Is your concern that you don't want to have people creating JSON schema file that might not be valid by creating this PR public?

That is one concern. And the other is, that it wouldn't have a use. That's why I've proposed adding it when it actually has a use.

> but I'm not sure why you want to have parameter generation feature available before having this guideline merged.

In [my previous post above](https://github.com/autowarefoundation/autoware-documentation/pull/367/files#r1205435380), I've stated

1. CI actions are in place to:
  - Validate the schema files (essential)
  - Validate the `.yaml` files with the schemas (essential)
  - Create `settings.json` files to support VS Code (desirable but not critical)
2. The `generate_parameter_library` can accommodate JSON Schemas, ensuring we do not end up with multiple sources of truth (desirable but not critical)

So, to me, parameter generation is **not** a necessity to add this.

@isamu-takagi said:
> For the time being, parameter definition file is a kind of README. Package maintainers write parameter definition file instead of writing the parameters section of README.

I looked into it to be more than just a Readme file. That's why I've raised my points.

If we are going to see it that way, I am also OK with this. Then I can just treat it like any other text file.

If there are no more objections, I can mark this conversation as resolved.

isamu-takagi replied:
>We have a list of the nodes which need to be updated, find the spreadsheet [here](https://docs.google.com/spreadsheets/d/1yAnXedY3zv2oD1prDZ_AyYiFMI8T9ZVrRNqm9kEPYcs/edit?usp=sharing) and look at the tab "ROS Node Config". @mitsudome-r has gone through the list to remove any ROS package which doesn't need to be modified.
>
>Edit: so to answer your question, I think anyone can make the changes but the package maintainer should probably review the changes to ensure it looks OK.

I understood @xmfcx -san's concern. So there are cases someone will introduce a JSON schema to the packages I maintain. It goes against my understanding. It is forced to create a definition file and its format is fixed.

To be acceptable, leave schema file maintenance to the schema author rather than the package maintainer. In other words, I don't update the schema file when I update my package's parameters. The author of the schema file will update it if necessary and I will review it. Then it's OK for me.

doganulus replied:
@isamu-takagi It is better that the node maintainer maintains the schema. No one else knows node parameters better than the node maintainer. I don't know the minimum and maximum values for your parameters or any other detail. 

JSON schema is written in the standard JSON and is widely used as an Internet standard with many benefits now or in the future. We have discussed this in great detail. Please see the previous discussions. This cannot be simpler than that. 

What @xmfcx mentions is about how to reap these benefits further. This doesn't bring an additional load to the maintainer. Just write the schema and enjoy benefits...

kminoda commented on docs/contributing/coding-guidelines/ros-nodes/parameters.md:
```diff
@@ -1,36 +1,150 @@
 # Parameters
```
As a reviewer, let me ask one very basic and high level question:

What is DevOps Dojo?
(The related links are bit too long for me to go through in a short period of time, so a short summary of its idea is very much appreciated, thanks!)

kaspermeck-arm replied:
@kminoda 

[DevOps Dojos](https://docs.google.com/presentation/d/1hsLgp1myeKhelRjOedA5yi-UK6ZNfwAUxRbvxdSdX_g/edit?usp=sharing) is the way we structure work in the Open AD Kit WG. Each dojo addresses a certain topic, in this case the ROS parameters.

Find our Kanban board [here](https://github.com/orgs/autowarefoundation/projects/6/views/4).

kminoda commented on docs/contributing/coding-guidelines/ros-nodes/parameters.md:
```diff
@@ -1,36 +1,150 @@
+- [About ROS 2 Parameters](https://docs.ros.org/en/humble/Concepts/About-ROS-2-Parameters.html)
 
-Autoware has the following two types of parameter files for ROS packages:
+## JSON Schema
```
As an developer who have not followed the AWF DevOps discussions, here are some high level comments on this modification:

1. Could you elaborate more on the background of this documentation? As a beginner, I could not get the whole idea until following the latest thread in this PR (especially [the comment by mitsudome-san](https://github.com/autowarefoundation/autoware-documentation/pull/367/files#r1205552883)). Maybe adding some phrase like "to have a single source of truth" helps.
2. Putting this section at the very top of this page is confusing. Consider moving this section to the last of this page or creating a separate page for this.

Comments are also welcome from @kaspermeck-arm and others as well!

kaspermeck-arm commented:
@xmfcx @mitsudome-r 

Thank you for your engagement on this topic!

I would like to propose the following:
- The PR is merged as is
  - The [new guidelines](https://autowarefoundation.github.io/autoware-documentation/pr-367/contributing/coding-guidelines/ros-nodes/parameters/) improve the [current guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/)
  - @mitsudome-r, once merged, please make changes you see appropriate to improve the guidelines
- @ambroise-arm and @kaspermeck-arm will start working on the GitHub Actions to validate the schemas and parameter files
  - https://github.com/autowarefoundation/autoware-github-actions/issues/232
  - Tips/ideas on how to do this from @xmfcx and @doganulus will be very helpful
- @HamburgDave reaches out to the community asking people to start adopting ROS packages to the new guidelines
  - Find list of ROS packages [here](https://docs.google.com/spreadsheets/d/1yAnXedY3zv2oD1prDZ_AyYiFMI8T9ZVrRNqm9kEPYcs/edit#gid=1801060430) under tab "ROS Node Config"
  - This is the reason why Open AD Kit needs the new coding guidelines to be live

Working in parallel will enable us to make progress faster! What do you think?

isamu-takagi commented:
>  It is better that the node maintainer maintains the schema. No one else knows node parameters better than the node maintainer. I don't know the minimum and maximum values for your parameters or any other detail for example.

@mitsudome-r @doganulus 
I know that is desirable. What is important to me is being able to write the parameter definitions for my packages in PickNik format. This means I update only PickNik format as documentation, not generate code with parameter_generation_library.

For me, PickNik format is more friendly and the JSON schema feels complicated. Requiring JSON schema updates is a disincentive to package maintenance. I am concerned about that because I think there are many ROS developers who feel the same way.

Please do not rule out the possibility of participation by developers who wish to use PickNik format. If you want JSON schema, introduce a mechanism to automatically convert from PickNik format. Then everything will work fine.

kaspermeck-arm commented:
> > It is better that the node maintainer maintains the schema. No one else knows node parameters better than the node maintainer. I don't know the minimum and maximum values for your parameters or any other detail for example.
> 
> @mitsudome-r @doganulus I know that is desirable. What is important to me is being able to write the parameter definitions for my packages in PickNik format. This means I update only PickNik format as documentation, not generate code with parameter_generation_library.
> 
> For me, PickNik format is more friendly and the JSON schema feels complicated. Requiring JSON schema updates is a disincentive to package maintenance. I am concerned about that because I think there are many ROS developers who feel the same way.
> 
> Please do not rule out the possibility of participation by developers who wish to use PickNik format. If you want JSON schema, introduce a mechanism to automatically convert from PickNik format. Then everything will work fine.

We don't want to discourage developers from creating a schema because it's complicated. I can create more educational content for developers to lower the threshold of creating a new schema. If there's interest I can host a short Tech Talk (which we can record) on how schemas are used. What do you think?

If the schema file is fully populated it contains all the necessary information which is needed for the PickNik format. I.e.,
- type
- default_value
- description
- validation
  - see [Built-In Validators](https://github.com/PickNikRobotics/generate_parameter_library#built-in-validators) for details

I would suggest that we go from schema --> yaml, and not the other way around. This is because the schema is required and PickNik's library is optional. Does this sound like an OK solution?

isamu-takagi commented:
@kaspermeck-arm 
I think it's the parameter definition that's required, not the JSON schema. The PickNik format also works perfectly as a parameter definition. And these can be convertible to each other. So either of the following is fine. I don't understand why you exclude the latter option.
- Depelover update schema, then yaml will be updated automatically.
- Depelover update yaml, then schema will be updated automatically.

kaspermeck-arm commented:
> @kaspermeck-arm I think it's the parameter definition that's required, not the JSON schema. The PickNik format also works perfectly as a parameter definition. And these can be convertible to each other. So either of the following is fine. I don't understand why you exclude the latter option.
> 
>     * Depelover update schema, then yaml will be updated automatically.
> 
>     * Depelover update yaml, then schema will be updated automatically.

@isamu-takagi 

I don't quite understand. The parameter definition file is the schema. The GitHub Action will use the schema to validate the parameter yaml file.

```mermaid
flowchart TD
    NodeSchema[Schema file: *.schema.json]
    ParameterFile[Parameter file: *.param.yaml]
    WebDocumentation[Web documentation table]

    NodeSchema -->|Validation| ParameterFile
    NodeSchema -->|Generate| WebDocumentation
```

I think involving the PickNik format is creating confusion here. We concluded that using this format is optional and therefore it is not part of the coding guidelines, as following the coding guidelines is mandatory.

The parameter yaml file doesn't contain all the information that the schema does, so it's not possible to go from the parameter yaml file to the schema.

I would recommend reading through the discussions we've had on this topic and why we chose to use JSON Schema. This decision has been agreed on. See:
* https://github.com/orgs/autowarefoundation/discussions/3288
* https://github.com/orgs/autowarefoundation/discussions/3325
* https://github.com/orgs/autowarefoundation/discussions/3371

isamu-takagi commented:
>The parameter yaml file doesn't contain all the information that the schema does, so it's not possible to go from the parameter yaml file to the schema.

@kaspermeck-arm Thank you. I didn't know this. I will read the details.

doganulus commented:
> I think it's the parameter definition that's required, not the JSON schema. The PickNik format also works perfectly as a parameter definition. And these can be convertible to each other. So either of the following is fine. I don't understand why you exclude the latter option.

@isamu-takagi I have explained several times in the previous discussions about Picknik format vs Schema. Below is my personal yet educated summary, but not binding the OpenAD kit group. 

The main problem: Picknik's format is a cluttered, underworked, and non-standard format that tries to do the very same thing --describing the structure of a config file-- with a popular Internet standard having common tooling support. In order to pick Picknik's format over an established format, it should have 10x features. It doesn't have even a quarter. This is why I suggest not using Picknik's format.

isamu-takagi commented:
OK. Since no progress has been made by arguing here, let's try JSON schema. But most of the advantages mentioned are not considered by package maintainers because they don't use the parameter definition file to do anything. So for them ease of writing is everything. If they don't feel the benefits of requiring schema maintenance, this rule may become a dead letter. It needs to monitor the progress for a while and take appropriate actions according to the situation.

doganulus commented:
> OK. Since no progress has been made by arguing here, let's try JSON schema. But most of the advantages mentioned are not considered by package maintainers because they don't use the parameter definition file to do anything. So for them ease of writing is everything. If they don't feel the benefits of requiring schema maintenance, this rule may become a dead letter. It needs to monitor the progress for a while and take appropriate actions according to the situation.

@isamu-takagi Well, package maintainers still have to ensure the quality of their packages with or without a parameter definition file. Parameters must be well-defined, correct, and updated with or without a parameter definition file. Without parameter definitions, you have to do that manually every single time. You need to check whether updated values are valid or not. Since this is a slow and cumbersome process, probably you would introduce an error at some commit, maybe it would be discovered late and you will lose lots of hours in debugging, or even worse it would lead to a crash.

Well, I would not take such risks if I were a package maintainer and embrace modern automated testing and verification methods. Not having well-defined specifications and automated testing is a very expensive gamble. I would never dare such a thing.

isamu-takagi commented:
@doganulus I totally agree with you about the need to define parameters. But it doesn't have to be JSON Schema. I also see the benefit of that, but it's only for developers of validation and documentation tools etc. I think that many developers want a simple format, and JSON Schema seems unnecessarily complex and disadvantageous. So using JSON Schema will force many developers to put in more effort than necessary for the benefit of some developers.

My point is that the guidelines that many developers complain about don't work. If we really want a parameter definition file, we should choose a format that the package maintainers actively update. Note that creating a parameter definition file is optional and package maintainers have the choice not to support it. If you accept that risk, I will no longer object.

doganulus commented:
Developers may look at the issue from a single angle. Here we need to look at this problem of paramter definition. Yes, you may not not like some certain aspects of JSON Schema, which I also have a few personal dislikes, but overall it is a widely-used standard and this has HUGE HUGE advantages compared to using any in-house alternative. 

Actually existing options are like this:

1) They just write a slightly less verbose JSON file. Developers need to write the specification of the in-house format they want to use, and implement the validator, add CI/CD files, and maintain it during the Autoware lifetime.

2) They just write a slightly more verbose JSON file. The rest is already available.

Do you think developers will pick the Option 1? 

I highly doubt it. Just let them know how much work must be done for the Option 1.

isamu-takagi commented:
> Developers just write a slightly less verbose YAML file. Developers also need to write the specification of the in-house format they want to use, and implement the validator, test it extensively, add CI/CD files, and maintain all during the Autoware lifetime.

@doganulus The developer of the former is different from the developer of the latter. Developers who write YAML files just use the validator tool. A similar mechanism is used in ROS message definitions. Many depelovers just write a simple definition like [common_interface](https://github.com/ros2/common_interfaces). Only some developers who maintain [rosidl](https://github.com/ros2/rosidl) receive the disadvantages you say, many developers developers can easily define messages.

isamu-takagi commented:
By the way, the following are all valid parameter files for my_node that has parameters foo and bar. How do I write this in my JSON Schema?

Apply parameters to all nodes.
```
/**:
  ros__parameters:
    foo: 1
    bar: 2
```

Apply parameters to my_node.
```
/my_node:
  ros__parameters:
    foo: 1
    bar: 2
```

Apply parameters to my_node under namespaces.
```
namespace1:
  namespace2:
    my_node:
      ros__parameters:
        foo: 1
        bar: 2
```


Apply parameters to my_node by two parameter files (e.g. common part and others)
```
my_node:
  ros__parameters:
    foo: 1
```

```
my_node:
  ros__parameters:
    bar: 2
```

doganulus commented:
@isamu-takagi See the schemas here for the examples above: https://github.com/doganulus/autoware-schema/

And one problem related is that I am not aware of the specification document for the ROS parameter files. The documentation is [scarce](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html) and [anecdotal](https://docs.ros.org/en/foxy/How-To-Guides/Parameters-YAML-files-migration-guide.html). Please tell me if such a document exists as I see many gaps when I looked at it carefully. For example, 
    - What is the maximum nesting level of namespaces in YAML? Unbounded or hard-capped? 
    - What are the allowed keywords for ROS parameters? For example, we see `ros__parameters` is a special keyword but where is it explicitly stated in the documentation? Are there any other such keywords? What is the reason to add additional nesting if not? 
    - Same for `/**`. Visible in the examples but what is it exactly? A regular expression or a glob? These are different things. I guess but this shouldn't be left to guesswork. Can I write `/**/my_node`? Probably but again this is a guesswork.
    - What are the rules for mapping cases in YAML files?

So, if any developer wants to develop a specification format, then they should answer such necessary questions but see that even core ROS modules fail to achieve this. That's why I keep telling the community to avoid half-baked formats and tooling. I consider it a serious safety problem.

isamu-takagi commented:
@doganulus 
Thanks for creating the sample files. But all of them are parameter files for the same node, so there should be one schema file. I didn't know how to write it.

@kaspermeck-arm 
My question is not resolved. As I commented earlier, how do I write a schema that supports remapped node name and divided parameter files? I have a parameter file that cannot use wildcards because I want to set different parameters for multiple nodes by one parameter file like this.

```
# parameters for foo node
/node1:
  ros__parameters:
    topic_rate: 10
    other_foo_node_param: ...

# parameters for bar node
/node2:
  ros__parameters:
    topic_rate: 5
    other_bar_node_param: ...
```

kaspermeck-arm commented:
@isamu-takagi 

Using the `joy_controller` as an example again, modifying the last section.
```JSON
  "properties": {
    "/node1": {
      "type": "object",
      "properties": {
        "ros__parameters": {
          "$ref": "#/definitions/joy_controller"
        }
      },
      "required": ["ros__parameters"]
    },
    "/node2": {
      "type": "object",
      "properties": {
        "ros__parameters": {
          "$ref": "#/definitions/joy_controller"
        }
      },
      "required": ["ros__parameters"]
    }
  },
  "required": ["/node1", "/node2"]
```

isamu-takagi commented:
@kaspermeck-arm
So I think it works well for node documentation. However, if you use it to validate node user’s parameter files, you will have problems. If your purpose is documentation, I think there is no problem for the time being.

doganulus commented:
> What node developers can define is the schema of the node's parameters (not a parameter file).

@isamu-takagi This is exactly what we ask from them. That's why we let them describe parameters under `definitions` schema key and a schema reference calls the definition. This is to validate `params.yaml` files in the package. If the developer does not remap, then it wouldn't need to deal with the cases you mention. The boilerplate coming from the unnecessary nesting of ROS parameters would be untouched by the developer. 

Further we can collect those definitions and merge them into a big schema file to be published online. Then advanced use cases as you described then can reference them in launch files. But first let developer define their local schema files.

isamu-takagi commented:
@doganulus Sorry, I made a mistake with the mention. Anyway, I'm relieved that the schema file doesn't prevent node user's parameter file structure.

isamu-takagi commented:
@kaspermeck-arm 
> When a developer is introduced to something which is new, in this case using JSON Schema, which is a cloud-native best-practice, we need to educate the community. This is growth and development.

Yes, you who need it have to achieve it. I've already commented to try it with JSON Schema, so it's fine to proceed with the PR.

> The PR is merged as is
> - The new guidelines improve the current guidelines
> - @mitsudome-r, once merged, please make changes you see appropriate to improve the guidelines

Please at least reflect [this comment](https://github.com/autowarefoundation/autoware-documentation/pull/367#discussion_r1205575336). The order of chapters is confusing. I think the natural order for developers is:
- Declare parameters
- Create parameter file
- Launch node with parameter file
- Create parameter schema file (JSON Schema is the means, use the purpose for chapter name)

kaspermeck-arm commented:
@isamu-takagi 

> 
> Please at least reflect [this comment](https://github.com/autowarefoundation/autoware-documentation/pull/367#discussion_r1205575336). The order of chapters is confusing. I think the natural order for developers is:
> 
>     * Declare parameters
>     * Create parameter file
>     * Launch node with parameter file
>     * Create parameter schema file (JSON Schema is the means, use the purpose for chapter name)
>

I suggest that we merge this PR as it is as it is a **big** change from the current coding guidelines. If you would like to change the order of the sections, create a PR with the proposed changes.

@xmfcx has agreed to this approach, see https://discord.com/channels/953808765935816715/953877118415151205/1113117316843044974.

Can we agree on this approach and merge this PR?

isamu-takagi commented:
>Please at least reflect https://github.com/autowarefoundation/autoware-documentation/pull/367#discussion_r1205575336. The order of chapters is confusing. I think the natural order for developers is:

@mitsudome-r @xmfcx At least two reviewers have pointed this out. Is there any reason you want to ignore this and merge? I can't understand why they would refuse to make a simple change, just swap the order of the chapters.

kaspermeck-arm commented:
@isamu-takagi 

Having the definition in place prior to creating the parameter file is also an option, there is nothing wrong with it, it's a preference. And I do want to highlight that the coding guidelines have already been approved.

I am not opposed to making changes to the coding guidelines once this PR has been merged. The purpose is to make it easier for the developers. Once merged, we can ask the community for feedback on the topic of usability and iterate on the coding guidelines until the community is satisfied with them.

isamu-takagi commented:
@kaspermeck-arm 
It is your job to reflect the reviewer's point out. Don't impose it on others.

isamu-takagi commented:
@kaspermeck-arm 
In fact, two reviewers pointed out that it was confusing, and it took no time at all to fix it. why not do that?

mitsudome-r commented:
@isamu-takagi If you are concerned to have a page that might be potentially confusing for some people, maybe I can create a PR to Kasper's documentation before merging his.

isamu-takagi commented:
@mitsudome-r I'm just saying follow the normal review process. According to Autoware's guidelines, it's the pull request owner's job to reflect the reviewer's comments. But if you specifically allow merging this PR, that's fine. I will update later. However, if this is accepted for everyone, the pull request guidelines will not work.

mitsudome-r commented:
@isamu-takagi 
We have discussed this in Open AD Kit call today with @kaspermeck-arm.
From the discussion, we have agreed that @kaspermeck-arm to make the following changes to address the feedback:
- change the order of sections as mentioned in this [comment](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1569417302.) 
- Add some details about rationale behind adding JSON schema file.

@kaspermeck-arm, for the second point, I would leave it up to you how you make the changes in the end, but here are some of my suggestions:
-  Perhaps using the images in this [comment](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1563641683) might be useful.
- optionally, listing reference links to past discussions might be also helpful for readers to track further details.

kaspermeck-arm commented:
@isamu-takagi @kminoda @mitsudome-r 

Thanks for the discussion and correspondence! I've made updates to the coding guidelines as was decided in the Open AD Kit WG meeting today (June 21st).

isamu-takagi commented:
@kaspermeck-arm Thank you for updating. The latest documentation makes it easier to understand the benefits of schema files. I wait for @mitsudome-r -san's approval.

xmfcx commented on docs/contributing/coding-guidelines/ros-nodes/parameters.md:
````diff
@@ -1,36 +1,175 @@
+      NodeSchema -->|Generate| WebDocumentation
+  ```
+
+*
````
The pre-commit ci is complaining about this line, i think it's a typo.

kaspermeck-arm replied:
Yes, it is, removing it!

taikitanaka3 commented:
@kaspermeck-arm 
Thank you for PR, but I think it's quite hard to see this PR for other maintainers too because
- There are a lot of diffs for one small yaml file [like this PR](https://github.com/autowarefoundation/autoware.universe/pull/3992).
- As @isamu-takagi -san [says](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1569417302), most of Maintainer doesn't know the correct format of this json schema.
- We don't know the way to validate values you added to this PR. May there should be some ways to test with CI.

kaspermeck-arm commented:
> @kaspermeck-arm Thank you for PR, but I think it's quite hard to see this PR for other maintainers too because
> 
>     * There are a lot of diffs for one small yaml file [like this PR](https://github.com/autowarefoundation/autoware.universe/pull/3992).
> 
>     * As @isamu-takagi -san [says](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1569417302), most of Maintainer doesn't know the correct format of this json schema.
> 
>     * We don't know the way to validate values you added to this PR. May there should be some ways to test with CI.

@taikitanaka3 - Since a few weeks ago, GitHub Actions have been put in place to validate the schema and the parameter file. Regarding the code review, what's expected from the owner is:
- ensure that the parameter description is accurate
- ensure that the bounds are correct

The rest should be handled by the CI and owner of the PR. 

How does this sound?

taikitanaka3 commented:
@kaspermeck-arm
Thank you for your comment I understand CI is included in github action but still two concerns left.

- There are a lot of diffs for one small yaml file https://github.com/autowarefoundation/autoware.universe/pull/3992.
- As @isamu-takagi -san https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1569417302, most of Maintainer doesn't know the correct format of this json schema.

doganulus commented:
@taikitanaka3 We proposed schemas in the OpenAD Kit with @kaspermeck-arm and others.

Schema is the description of the parameter file. You explain the parameter file to a computer. What is the type, what are allowed values, what is its structure? Of course, it will be longer than the instance. I think your first concern is a misunderstanding of the purpose of the schema files.

The second is more about safety culture. Previously parameter files have been unchecked, uncontrolled, and untested. This is a safety problem. Think the schema is a safety belt in your vehicle. No one likes a safety belt until there happens a crash, and it saves your life. The infamous example of the [Ariane5](https://en.wikipedia.org/wiki/Ariane_flight_V88) rocket crashed in 1996 because of a single type error in their software. Maintainers are responsible for not letting this happen for Autoware. If we tell them in that way, they would immediately start to learn the correct format and every other safety feature we will introduce.

kaspermeck-arm commented:
> @kaspermeck-arm Thank you for your comment I understand CI is included in github action but still two concerns left.
> 
>     * There are a lot of diffs for one small yaml file [refactor(joy_controller): rework parameters autoware.universe#3992](https://github.com/autowarefoundation/autoware.universe/pull/3992).
> 
>     * As @isamu-takagi -san [docs(parameters): add DevOps Dojo: ROS Node Config documentation #367 (comment)](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1569417302), most of Maintainer doesn't know the correct format of this json schema.

@taikitanaka3 

I thought I addressed your two concerns in [my comment](https://github.com/autowarefoundation/autoware-documentation/pull/367#issuecomment-1643862815) above.

To me this sounds like you and @isamu-takagi would benefit from understanding schemas better. I would recommend exploring https://json-schema.org/learn/ which has very useful information about `JSON Schema`. In short, a schema is meta-data, that is, data about data, creating the structure for the parameter files. Let me know if there's anything the Open AD Kit WG can do for you to help assist your learning!
